docs/topics/coroutines.rst

```diff
@@ -103,3 +108,68 @@ Common use cases for asynchronous code include:
 
 .. _aio-libs: https://github.com/aio-libs
+
+.. _async-spider-middlewares:
+
+Asynchronous spider middlewares
+===============================
+
+.. versionadded:: VERSION
+.. note:: This currently applies to
+          :meth:`~scrapy.spidermiddlewares.SpiderMiddleware.process_spider_output`.
+
+Middleware methods discussed here can take and return async iterables. They can
+return the same type of iterable or they can take a normal one and return an
+async one. If such method needs to return an async iterable it must be an async
+generator, not just a coroutine that returns an iterable.
+
+.. autofunction:: scrapy.utils.asyncgen.as_async_generator
+
+In the simplest form that supports both sync and async input it can be written
+like this::
+
+    from scrapy.utils.asyncgen import as_async_generator
+
+    class ProcessSpiderOutputAsyncGenMiddleware:
+        async def process_spider_output(self, response, result, spider):
+            async for r in as_async_generator(result):
+                # ... do something with r
+                yield r
+
+If the middleware input (the callback result for ``process_spider_output``) is
+an async iterable, all middlewares that process it must support it. The
+built-in ones do, but the ones in your project and 3rd-party ones will need to
+be updated to support it, as the code that expects a normal iterable will break
+on an async one. If these middlewares receive an async iterable, they must
+return one as well. On the other hand, if they receive a normal iterable, they
+shouldn't break and ideally should return a normal iterable too. There can be
+several possible implementations of this.
+
+The simplest one, always converting normal iterables to async ones, is provided
+above. Because a result of a middleware method is passed to the same method of
+the next middleware, it's only possible to mix middlewares with synchronous and
+asynchronous implementations of the same method if all synchronous ones are
+called first (which isn't always possible).
+
+Another option is to make separate methods for normal and async iterables and
+choose one at run time::
+
+    from inspect import isasyncgen
+
+    class ProcessSpiderOutputAsyncGenMiddleware:
+        def _normal_process_spider_output(self, response, result, spider):
+            # ... do something with normal result
+
+        async def _async_process_spider_output(self, response, result, spider):
+            # ... do the same with async result
+
+        def process_spider_output(self, response, result, spider):
+            if isasyncgen(result):
+                return self._async_process_spider_output(self, response, result, spider)
+            else:
+                return self._normal_process_spider_output(self, response, result, spider)
+
+If you are writing a middleware that you intend to publish or to use in many
+projects, this is likely the best way to implement it. It may be possible to
```

Comment:
I wonder if we should be thinking about eventually deprecating non-async process_spider_output. Without doing it, users will always need to write code like the above, which complicates middleware implementation. They'd also need to call as_async_generator always, as `result` can be of different type.

If we disregard backwards compatibility, and all Scrapy ecosystem is updated, won't 

```py
    class ProcessSpiderOutputAsyncGenMiddleware:
        async def process_spider_output(self, response, result, spider):
            async for res in result:
                # ... do something with res
```

be a better API? If so, is there a plan to eventually get there?

Reply:
I’ve been thinking that Scrapy 3.0 could be about this, dropping support for synchronous definitions where async is supported. I was thinking of the expected behavioral change for `start_requests`, not this, but the point is the same. So +1 from me.

Reply:
I agree that making everything async will greatly simplify the code, but I'm not sure what should be the steps to that.

Normally there are 4-5 parts of middleware-related code that can be sync, async or both:

1. Spider methods (depends on the specific middleware method discussed).
2. Methods of middlewares in the user project.
3. Methods of 3rd-party middlewares used in the user project.
4. Methods of bundled middlewares.
5. The Scrapy code that runs middleware code on the spider method results.

4 and 5 can be rewritten when the Scrapy maintainers decide, 1 and 2 can be rewritten by the user when they decide or when it's required by other components, but I'm not sure what the maintainers of 3 should do. Initially the code should be universal but at which point can they convert it to async-only? After the Scrapy maintainers declare the sync support deprecated but before they remove it from 4 and 5?

Reply:
> at which point can they convert it to async-only? After the Scrapy maintainers declare the sync support deprecated but before they remove it from 4 and 5?

It’s a good point, their code will stop working with the latest Scrapy if they have not made the switch by the time Scrapy has. 

Besides increasing the Scrapy major version to make this aspect obvious, we could consider increasing the time between the original deprecation to the support removal for this specific case, e.g. 2 years instead of 1 year. So, if the last of this deprecations is introduced in Scrapy 2.n, we would not release Scrapy 3.0 until that time after the release of 2.n.

On the other hand, 1 year may already be enough, and if those middlewares do not upgrade users can always keep Scrapy in a 2.x version, as as people that still need Python 2 use Scrapy 1.8. Similarly, I imagine we can keep backporting security fixes to 2.x after 3.0 is released.

Reply:
We discussed it on a call. Thoughts:

1. It'd be nice not to require third-party and user middlewares to maintain compatibility code.
2. it seems this can be achieved by auto-promoting sync to async, and auto-downgrading async to sync. We should be warning when downgrading happens. Warning for auto-promoting is not required, because it is a safe operation. We can add it later if it proves problematic.
3. In this case the idea is to suggest always writing `async def process_spider_output`. API would change between sync and async versions: in async def version `results` is going to be async iterator, not a sync one. It means there is no reason to use as_async_generator in the user code, simplifying the API.

@wRAR is to try to implement this approach.

Reply:
A comment I've been writing for some time. I think it is not valid, but don't recall why :)

---

@wRAR a good question. 

If we rewrite them as async def, it'd mean all regular middlewares would start to issue warnings, right?

If we maintain compatibility code, it'd mean warnings would be only issued when there is `async def parse` method, or when there is some async def middleware.

Let's say we choose to rewrite built-in middlewares to async def. In this case users would be getting warnings after updating to latest Scrapy version. Third-party middlewares would need to be updated. But the issue is that third-party middlewares would want to support older Scrapy versions as well, where async def won't work (even with warnings). It means they'd need to add compatibility code, which is exactly what we'd like to avoid.

So, it seems it is better to keep built-in middlewares universal.

It'd mean that there won't be warnings if all third-party and user middlewares as sync, but as soon as async def parse is used, or an async def middleware is inserted in a project which also has sync middlewares, warnings would start to appear. That's not ideal, but hopefully it'd give users some time.

When talking about compatibility, there are actually 2 concerns:

1) compatibility with different mixes of def and async def code,
2) compatibility with older Scrapy versions

Reply:
@kmike You do mention that we want to avoid third-party developers having to make their spider middlewares universal. However, to maintain support with old Scrapy versions, I imagine that’s the only possible route.

Moreover, universal implementations support coroutines while avoiding async-to-sync conversions. Those conversions are problematic (I do not recall why, we should cover this in the documentation), so they generate a warning. Avoiding that warning is argument enough I think for both built-in and third-party middlewares to aim for universal implementations.

So I expect third-party middlewares to go for ugly, universal implementations, until they decide to drop support for Scrapy 2.5 and earlier. While not ideal, I cannot see any other way to go about this.

I think the best we can do to improve the experience is to try an make it as easy as possible to implement these universal spider middlewares. Although I’m concerned any step in that direction would make things more messy in the long term.

We cannot provide utility code to make them easier to write, because it would not be available in old Scrapy versions.

The only thing I think we could do is allow avoiding the `isasyncgen` call by allowing to define a coroutine implementation with a different method name (e.g. async_process_spider_output) which old Scrapy versions would simply ignore. That’s the best we can do, I think. But I’m not even sure it is worth the trouble of later deprecating that method.